File: src/main/java/uss/code/registration/infra/RegistrationTypeResolver.java
package uss.code.registration.infra;

import lombok.experimental.UtilityClass;
import uss.code.course.domain.Course;
import uss.code.course.domain.CourseDepartment;
import uss.code.member.domain.Member;

import java.util.Set;

import static uss.code.course.domain.CourseClassification.GENERAL_ELECTIVE;
import static uss.code.course.domain.CourseClassification.MAJOR_ADVANCED;
import static uss.code.course.domain.CourseClassification.MAJOR_BASIC;
import static uss.code.course.domain.CourseClassification.MAJOR_CORE;

@UtilityClass
public class RegistrationTypeResolver {

    private static final String GENERAL_ELECTIVE_NAME = GENERAL_ELECTIVE.getName();
    private static final Set<String> MAJOR_CLASSIFICATION_CODES = Set.of(
            MAJOR_BASIC.getCode(), MAJOR_CORE.getCode(), MAJOR_ADVANCED.getCode()
    );

    public static String resolve(
            final Member member,
            final Course course
    ) {
        if (isOtherDepartmentMajor(member, course)) {
            return GENERAL_ELECTIVE_NAME;
        }

        return course.getClassificationName();
    }

    private static boolean isOtherDepartmentMajor(
            final Member member,
            final Course course
    ) {
        final boolean ownDepartment = CourseDepartment.ownedBy(member.getDepartment())
                .contains(course.getDepartment());

        return !ownDepartment && MAJOR_CLASSIFICATION_CODES.contains(course.getClassificationCode());
    }
}

File: src/main/kotlin/uss/code/registration/infra/RegistrationTypeResolver.kt
package uss.code.registration.infra

import uss.code.course.domain.Course
import uss.code.course.domain.CourseClassification.GENERAL_ELECTIVE
import uss.code.course.domain.CourseClassification.MAJOR_ADVANCED
import uss.code.course.domain.CourseClassification.MAJOR_BASIC
import uss.code.course.domain.CourseClassification.MAJOR_CORE
import uss.code.course.domain.CourseDepartment
import uss.code.member.domain.Member

object RegistrationTypeResolver {
    private val GENERAL_ELECTIVE_NAME = GENERAL_ELECTIVE.displayName
    private val MAJOR_CLASSIFICATION_CODES = setOf(MAJOR_BASIC.code, MAJOR_CORE.code, MAJOR_ADVANCED.code)

    fun resolve(
        member: Member,
        course: Course,
    ): String {
        if (isOtherDepartmentMajor(member, course)) {
            return GENERAL_ELECTIVE_NAME
        }

        return course.classificationName
    }

    private fun isOtherDepartmentMajor(
        member: Member,
        course: Course,
    ): Boolean {
        val ownDepartment = course.department in CourseDepartment.ownedBy(member.department)

        return !ownDepartment && course.classificationCode in MAJOR_CLASSIFICATION_CODES
    }
}
